File: src/edu/oswego/csc436/states/Leader.java
package edu.oswego.csc436.states;

import static edu.oswego.csc436.Constants.*;
import edu.oswego.csc436.data.BadSensorValueException;
import edu.oswego.csc436.data.Data;

import java.io.IOException;

public class Leader extends State {
  private static Leader instance = new Leader();

  @Override
  public State update(float dt, Data data) {
    try {
      data = data.update(dt, Math.min(SPEED_LIMIT, data.getTargetSpeed() + dt * ACCELERATION_RATE));
      data.writeSpeeds();

      if (data.objectDetected()) {
        if (data.getUssValue() < CRITICAL_DISTANCE) {
          return Emergency.getInstance();
        } else if (data.getUssValue() < STOP_DISTANCE) {
          return Stop.getInstance();
        } else if (data.getUssValue() < FOLLOW_DISTANCE) {
          return Follow.getInstance();
        }
      }
      return this;
    } catch (BadSensorValueException | IOException exception) {
      exception.printStackTrace(System.err);
      return this;
    }
  }

  public static State getInstance() {
    return instance;
  }
}

File: src/state/leader.py
import sys

from constants import Constants
from data import BadSensorValueError
from state import State
import emergency
import follow
import stop
import gopigo.gopigo

class Leader(State):
    _instance = None

    def __init__(self):
        pass

    def update(self, dt, data):
        try:
            gopigo.fwd()
            if data.object_detected():
                uss_value = data.get_uss_value()
                if uss_value < data.dynamic_critical_distance():
                    return emergency.Emergency.get_instance()
                elif uss_value < data.dynamic_stop_distance():
                    return stop.Stop.get_instance()
                elif uss_value < Constants.FOLLOW_DISTANCE:
                    return follow.Follow.get_instance()

            target_speed = data.get_target_speed()
            if target_speed < Constants.SPEED_LIMIT:
                data.set_speed(min(target_speed + (Constants.ACCELERATION_RATE * dt), Constants.SPEED_LIMIT))
            else:
                data.set_speed(max(target_speed - (Constants.ACCELERATION_RATE * dt), 0.0))

            return self

        except (IOError, BadSensorValueError) as e:
            sys.stderr.write(e)
            return self

    @staticmethod
    def get_instance():
        if Leader._instance == None:
            Leader._instance = Leader()
        return Leader._instance
